### DISCUSSION TITLE
Model time limit issue

### DISCUSSION THREAD

**Discussion by sakuraIsNow** [spacepy/spacepy#801] [Support and Q&A]
I have replaced the model with igrf14, but I still encounter the following warning. How can I resolve it?

*** WARNING -- Input year = 2026.50
 is out of valid range 1900-2025 Using nearest ***

**Comment by drsteve** [comment]
Great question. The short answer is that the warning you're seeing is from the IRBEM library, which SpacePy wraps. So updating to IGRF14 for coordinate conversions is trivial as it's just replacing a coefficient file (instructions below for future visitors). Anything needing the IRBEM library (e.g., `spacepy.irbempy.get_Lm`) would need the (Fortran77) IRBEM library to be updated as it doesn't use a standard coefficient file. The longer, and more useful, answer is below.


For reference (for those who may be looking to update to IGRF14), SpacePy uses the standard format IGRF coefficient file that's available from https://www.ngdc.noaa.gov/IAGA/vmod/coeffs/igrf14coeffs.txt
While currently the IGRF13 file is bundled with SpacePy, the easy way to update is to copy the IGRF14 file to `.spacepy/data/igrfcoeffs.txt` (or have separate `igrf13coeffs.txt` and `igrf14coeffs.txt` files and copy/symlink whichever you want to use to `igrfcoeffs.txt`.)

So once that's done, _**anything using the SpacePy implementation of IGRF**_ will work (without warnings) through the valid span of IGRF14. So that covers the `igrf` module and coordinate conversion with the default SpacePy backend.

_**Anything using the IRBEM library**_ -- things from `spacepy.irbempy` and coordinate transforms using `use_irbem=True` -- will use the IRBEM implementation of IGRF (because we're wrapping a 3rd-party FORTRAN library to access their magnetic field models, etc.) IRBEMlib doesn't use the same coefficient files - it has the coefficients hardcoded in `igrf_coef.f` (see https://github.com/spacepy/spacepy/blob/main/spacepy/irbempy/irbem-lib-20220829-dfb9d26/source/igrf_coef.f)
So to have full IGRF14 support, that also needs updating. While we can update locally, it would be ideal to have the update at https://github.com/PRBEM/IRBEM (see https://github.com/PRBEM/IRBEM/blob/main/source/igrf_coef.f)and then pull the latest IRBEM into SpacePy.

So, to illustrate:

First let's open ipython and import the `spacepy.igrf` module. Note that most users won't ever use this directly as it's a bare-bones implementation that's used for defining the dipole axis and magnetic moment (currently, I'm planning a full model implementation but need time to write it)
```
(default310) :~$ ipython
Python 3.10.5 | packaged by conda-forge | (main, Jun 14 2022, 07:04:59) [GCC 10.3.0]
Type 'copyright', 'credits' or 'license' for more information
IPython 8.4.0 -- An enhanced Interactive Python. Type '?' for help.

In [1]: import datetime as dt
   ...: from importlib import reload
   ...: import os
   ...: import shutil
   ...: import spacepy
   ...: from spacepy import igrf
   ...: import spacepy.coordinates as spc
   ...: import spacepy.time as spt
This unreleased version of SpacePy is not supported by the SpacePy team.
```
Now we can check moments and validity of a requested date. With the packaged IGRF13, we should get an out-of-range warning for 2028.
```
In [2]: imod = igrf.IGRF(); imod.initialize(dt.datetime(2028,4,10)); imod.calcDipoleAxis()
   ...: imod.moment
<ipython-input-2-3a03befac466>:1: UserWarning: IGRF: Requested time is outside valid range.
Valid range is [1900-01-01 00:00:00, 2025-01-01 00:00:00]
Proceeding using effective date 2025-01-01 00:00:00
  imod = igrf.IGRF(); imod.initialize(dt.datetime(2028,4,10)); imod.calcDipoleAxis()
Out[2]: {'cd': 29756.048665517297, 'cd_McIlwain': 31165.3, 'cd_2010': 29950.126496}
```
And now let's replace the coefficient file with IGRF14 and try again.
```
In [3]: coeff13 = os.path.join(spacepy.DOT_FLN, 'data', 'igrf13coeffs.txt')
   ...: coeff14 = os.path.join(spacepy.DOT_FLN, 'data', 'igrf14coeffs.txt')
   ...: coeff_use = os.path.join(spacepy.DOT_FLN, 'data', 'igrfcoeffs.txt')
   ...: shutil.copy(coeff14, coeff_use)
   ...: reload(igrf)
   ...: imod = igrf.IGRF(); imod.initialize(dt.datetime(2028,4,10)); imod.calcDipoleAxis()
   ...: imod.moment
Out[3]: {'cd': 29680.450338789367, 'cd_McIlwain': 31165.3, 'cd_2010': 29950.126496}
```
Note that we now have no warning and the strength of the centered dipole has changed for the requested date.
This is now using IGRF14.

So let's test some coordinate transforms with IGRF14.
```
In [4]: cc_spacepy = spc.Coords([3,2,1], 'GSM', 'car', use_irbem=False)
   ...: cc_irbem = spc.Coords([3,2,1], 'GSM', 'car', use_irbem=True)
   ...: cc_spacepy.ticks = spt.Ticktock('2024-06-30')
   ...: cc_irbem.ticks = spt.Ticktock('2024-06-30')
   ...: cc_spacepy.convert('GEO', 'car')
Out[4]: Coords( [[-2.465238365186453, -2.1765140042256395, 1.7847651364318224]] , 'GEO', 'car')

In [5]: cc_irbem.convert('GEO', 'car')
Out[5]: Coords( [[-2.4640924275808636, -2.1765641153345534, 1.786285856232579]] , 'GEO', 'car')
```
As you can see the results for 2024 are very similar. There are two reasons for the differences: 1. a subtle difference in definition/implementation of GSM, and 2. different IGRF coefficients (IRBEMlib is using 13 but we've updated SpacePy to use 14).

So now we can move out of range for IGRF13:
```
In [6]: cc_spacepy.ticks = spt.Ticktock('2026-06-30')
   ...: cc_irbem.ticks = spt.Ticktock('2026-06-30')
   ...: cc_spacepy.convert('GEO', 'car')
Out[6]: Coords( [[-2.464129547576557, -2.174596158541711, 1.7886299561440255]] , 'GEO', 'car')

In [7]: cc_irbem.convert('GEO', 'car')


 *** WARNING -- Input year = 2026.50
 is out of valid range 1900-2025 Using nearest ***


Out[7]: Coords( [[-2.4626035496393976, -2.17439756164635, 1.7909715243994937]] , 'GEO', 'car')
```
Here we see the conversion with the SpacePy backend work without warning, but the conversion with the IRBEM backend warns us we're out-of-range. And note the formatting of the warning here is different to the Python warning we saw above when we used `spacepy.igrf` outside the valid range. We can see that again by checking the (SpacePy) coordinate transform outside the valid range of IGRF14.
```
In [8]: cc_spacepy.ticks = spt.Ticktock('2036-06-30')
   ...: cc_spacepy.convert('GEO', 'car')
/home/me/miniconda3/envs/default310/lib/python3.10/site-packages/spacepy/ctrans/__init__.py:602: UserWarning: IGRF: Requested time is outside valid range.
Valid range is [1900-01-01 00:00:00, 2030-01-01 00:00:00]
Proceeding using effective date 2030-01-01 00:00:00
  self['IGRF'].initialize(self['UTC'].UTC[0])
Out[8]: Coords( [[-2.4623335214997444, -2.1741297189808475, 1.7916678246658706]] , 'GEO', 'car')
```

I'm planning to update the default IGRF in SpacePy to IGRF14 early in the new year. That includes pulling the latest IRBEM, but there are a lot of peripheral tasks around updating the wrapped Fortran library so it may take some time. An untested, short-term fix would be to just build SpacePy from source after copying https://github.com/PRBEM/IRBEM/blob/main/source/igrf_coef.f into the `spacepy/irbempy/irbem-lib-20220829-dfb9d26/source` directory.

**Reply by sakuraIsNow** [reply]
I am currently using `irbempy.get_Lm` to calculate L values, with the time set to 2026. Even after replacing the model with IGRF14, I still receive a warning about the time range. Additionally, the L values calculated before and after replacing the IGRF14 model remain unchanged.
